File: backend/app/fifo.py
"""Realized-lot matching engines.

realized_lots is a fully derived table: every SELL is matched against prior
BUYs of the same instrument. Two strategies are supported:

- ``fifo``: each SELL consumes BUY lots first-in-first-out, with each match
  recording the cost basis from that specific lot.
- ``average``: each SELL is costed at the running weighted-average cost of the
  open position; one realized row per SELL. The realized lot's ``open_date`` is
  the earliest still-open BUY date at sell time, so the holding period stays
  defensible.

Call ``recompute_realized_lots(db, method)`` after any change to the
transactions ledger to keep realized gains / tax reporting in sync.
"""

from collections import deque
from decimal import Decimal

from sqlalchemy import select
from sqlalchemy.orm import Session

from . import models

_LONG_TERM_DAYS = 365
_CENT = Decimal("0.01")


def _d(value) -> Decimal:
    return Decimal(str(value)) if value is not None else Decimal("0")


def recompute_realized_lots(db: Session, method: str = "fifo") -> int:
    """Rebuild realized_lots from the transactions ledger using ``method``."""
    markets = dict(
        db.execute(select(models.Instrument.id, models.Instrument.market)).all()
    )
    txs = db.scalars(
        select(models.Transaction)
        .where(models.Transaction.type.in_(("BUY", "SELL")))
        .order_by(models.Transaction.trade_date, models.Transaction.id)
    ).all()

    if method == "average":
        realized = _match_average(txs, markets)
    elif method == "fifo":
        realized = _match_fifo(txs, markets)
    else:
        raise ValueError(f"Unknown matching method: {method!r}")

    db.query(models.RealizedLot).delete(synchronize_session=False)
    db.add_all(realized)
    db.commit()
    return len(realized)


def _match_fifo(txs, markets):
    open_lots: dict[int, deque] = {}
    realized: list[models.RealizedLot] = []

    for tx in txs:
        qty = _d(tx.qty)
        if qty <= 0:
            continue
        unit_native = _d(tx.price_usd) * _d(tx.fx_rate)  # MXN per unit, pre-fee
        fee_per_unit = _d(tx.fees_mxn) / qty
        lots = open_lots.setdefault(tx.instrument_id, deque())

        if tx.type == "BUY":
            # Fees on a buy add to cost basis.
            lots.append({
                "tx_id": tx.id,
                "date": tx.trade_date,
                "qty": qty,
                "unit_cost": unit_native + fee_per_unit,
            })
            continue

        # SELL — fees on a sell reduce proceeds.
        unit_proceeds = unit_native - fee_per_unit
        remaining = qty
        while remaining > 0 and lots:
            lot = lots[0]
            matched = min(remaining, lot["qty"])
            cost = (matched * lot["unit_cost"]).quantize(_CENT)
            proceeds = (matched * unit_proceeds).quantize(_CENT)
            days = (tx.trade_date - lot["date"]).days
            realized.append(models.RealizedLot(
                instrument_id=tx.instrument_id,
                open_tx_id=lot["tx_id"],
                close_tx_id=tx.id,
                open_date=lot["date"],
                close_date=tx.trade_date,
                qty=matched,
                proceeds_mxn=proceeds,
                cost_mxn=cost,
                gain_mxn=proceeds - cost,
                holding_days=days,
                kind="long" if days > _LONG_TERM_DAYS else "short",
                market=markets.get(tx.instrument_id, "foreign"),
            ))
            lot["qty"] -= matched
            remaining -= matched
            if lot["qty"] <= 0:
                lots.popleft()
        # remaining > 0 here means an oversold position; unmatched qty is skipped.

    return realized


def _match_average(txs, markets):
    # Per instrument: a FIFO queue of open BUYs (just for date tracking), plus
    # running totals (qty + cost MXN) used to compute the weighted average.
    open_buys: dict[int, deque] = {}
    totals: dict[int, dict[str, Decimal]] = {}
    realized: list[models.RealizedLot] = []

    for tx in txs:
        qty = _d(tx.qty)
        if qty <= 0:
            continue
        unit_native = _d(tx.price_usd) * _d(tx.fx_rate)

        if tx.type == "BUY":
            buy_cost = qty * unit_native + _d(tx.fees_mxn)  # fees add to cost basis
            open_buys.setdefault(tx.instrument_id, deque()).append(
                {"date": tx.trade_date, "qty": qty}
            )
            t = totals.setdefault(
                tx.instrument_id, {"qty": Decimal("0"), "cost": Decimal("0")}
            )
            t["qty"] += qty
            t["cost"] += buy_cost
            continue

        # SELL — costed at the running average; oversold qty is skipped.
        t = totals.setdefault(
            tx.instrument_id, {"qty": Decimal("0"), "cost": Decimal("0")}
        )
        if t["qty"] <= 0:
            continue
        sell_qty = min(qty, t["qty"])
        avg = t["cost"] / t["qty"]
        cost_basis = (sell_qty * avg).quantize(_CENT)
        unit_proceeds = unit_native - _d(tx.fees_mxn) / qty
        proceeds = (sell_qty * unit_proceeds).quantize(_CENT)

        buys = open_buys.setdefault(tx.instrument_id, deque())
        open_date = buys[0]["date"] if buys else tx.trade_date
        # Advance the queue so the next sell's open_date reflects what's still held.
        remaining = sell_qty
        while remaining > 0 and buys:
            front = buys[0]
            matched = min(remaining, front["qty"])
            front["qty"] -= matched
            remaining -= matched
            if front["qty"] <= 0:
                buys.popleft()

        days = (tx.trade_date - open_date).days
        realized.append(models.RealizedLot(
            instrument_id=tx.instrument_id,
            open_tx_id=None,  # no single opening transaction under average cost
            close_tx_id=tx.id,
            open_date=open_date,
            close_date=tx.trade_date,
            qty=sell_qty,
            proceeds_mxn=proceeds,
            cost_mxn=cost_basis,
            gain_mxn=proceeds - cost_basis,
            holding_days=days,
            kind="long" if days > _LONG_TERM_DAYS else "short",
            market=markets.get(tx.instrument_id, "foreign"),
        ))
        t["qty"] -= sell_qty
        t["cost"] -= cost_basis
        if t["qty"] <= 0:
            t["cost"] = Decimal("0")  # avoid carrying a residual when position closes

    return realized

File: frontend/src/lib/fifo.js
/* Offline realized-lot matching engines — JS port of backend/app/fifo.py.
 *
 * Two strategies are supported via the `method` parameter:
 *  - "fifo"    (default): each SELL consumes BUY lots first-in-first-out.
 *  - "average": each SELL is costed at the running weighted-average cost; the
 *               realized lot's open_date is the earliest still-open BUY date.
 *
 * In live mode the backend computes these. When the backend is unreachable,
 * these helpers do the same client-side so adding / editing a transaction or
 * switching method still updates realized gains and the tax report. */

import { POSITIONS_RAW } from "./demoData.js";

const LONG_TERM_DAYS = 365;

// Mirrors Instrument.market on the backend (BMV listings are domestic).
const MARKET_BY_TICKER = Object.fromEntries(
  POSITIONS_RAW.map(p => [p.ticker, p.exchange === "BMV" ? "domestic" : "foreign"]),
);

const round2 = (n) => Math.round(n * 100) / 100;
const marketOf = (ticker) => MARKET_BY_TICKER[ticker] ?? "foreign";

function txNum(id) {
  const m = typeof id === "string" && id.match(/(\d+)/);
  return m ? Number(m[1]) : 0;
}

function daysBetween(openDate, closeDate) {
  return Math.round((new Date(closeDate) - new Date(openDate)) / 86400000);
}

function sortedBuySell(transactions) {
  return [...transactions]
    .filter(t => t.type === "BUY" || t.type === "SELL")
    .sort((a, b) => (a.date < b.date ? -1 : a.date > b.date ? 1 : txNum(a.id) - txNum(b.id)));
}

function realizedRow(tx, openDate, qty, costMxn, proceedsMxn) {
  const days = daysBetween(openDate, tx.date);
  return {
    closeDate: tx.date,
    openDate,
    ticker: tx.ticker,
    qty,
    proceedsMXN: proceedsMxn,
    costMXN: costMxn,
    gainMXN: round2(proceedsMxn - costMxn),
    days,
    kind: days > LONG_TERM_DAYS ? "long" : "short",
    market: marketOf(tx.ticker),
  };
}

/* Build realized lots from a transactions array.
 * `method` is "fifo" (default) or "average". */
export function computeRealizedLots(transactions, method = "fifo") {
  const txs = sortedBuySell(transactions);
  if (method === "average") return matchAverage(txs);
  if (method === "fifo")    return matchFifo(txs);
  throw new Error(`Unknown matching method: ${method}`);
}

function matchFifo(txs) {
  const openLots = new Map(); // ticker -> queue of { date, qty, unitCost }
  const realized = [];

  for (const tx of txs) {
    const qty = Number(tx.qty);
    if (!(qty > 0)) continue;
    const unitNative = Number(tx.priceUSD) * Number(tx.fxRate); // MXN per unit, pre-fee
    const feePerUnit = Number(tx.feesMXN) / qty;
    if (!openLots.has(tx.ticker)) openLots.set(tx.ticker, []);
    const lots = openLots.get(tx.ticker);

    if (tx.type === "BUY") {
      // Fees on a buy add to cost basis.
      lots.push({ date: tx.date, qty, unitCost: unitNative + feePerUnit });
      continue;
    }

    // SELL — fees on a sell reduce proceeds.
    const unitProceeds = unitNative - feePerUnit;
    let remaining = qty;
    while (remaining > 0 && lots.length > 0) {
      const lot = lots[0];
      const matched = Math.min(remaining, lot.qty);
      const cost = round2(matched * lot.unitCost);
      const proceeds = round2(matched * unitProceeds);
      realized.push(realizedRow(tx, lot.date, matched, cost, proceeds));
      lot.qty -= matched;
      remaining -= matched;
      if (lot.qty <= 0) lots.shift();
    }
    // remaining > 0 means an oversold position; unmatched qty is skipped.
  }
  return realized;
}

function matchAverage(txs) {
  const openBuys = new Map(); // ticker -> queue of { date, qty }  (for open_date tracking)
  const totals = new Map();   // ticker -> { qty, cost } running totals in MXN
  const realized = [];

  for (const tx of txs) {
    const qty = Number(tx.qty);
    if (!(qty > 0)) continue;
    const unitNative = Number(tx.priceUSD) * Number(tx.fxRate);

    if (tx.type === "BUY") {
      const buyCost = qty * unitNative + Number(tx.feesMXN); // fees add to cost basis
      if (!openBuys.has(tx.ticker)) openBuys.set(tx.ticker, []);
      openBuys.get(tx.ticker).push({ date: tx.date, qty });
      const t = totals.get(tx.ticker) ?? { qty: 0, cost: 0 };
      t.qty += qty;
      t.cost += buyCost;
      totals.set(tx.ticker, t);
      continue;
    }

    // SELL — costed at the running average; oversold qty is skipped.
    const t = totals.get(tx.ticker);
    if (!t || t.qty <= 0) continue;
    const sellQty = Math.min(qty, t.qty);
    const avg = t.cost / t.qty;
    const costBasis = round2(sellQty * avg);
    const unitProceeds = unitNative - Number(tx.feesMXN) / qty;
    const proceeds = round2(sellQty * unitProceeds);

    const buys = openBuys.get(tx.ticker) ?? [];
    const openDate = buys.length > 0 ? buys[0].date : tx.date;
    // Advance the queue so the next sell's open_date reflects what's still held.
    let remaining = sellQty;
    while (remaining > 0 && buys.length > 0) {
      const front = buys[0];
      const matched = Math.min(remaining, front.qty);
      front.qty -= matched;
      remaining -= matched;
      if (front.qty <= 0) buys.shift();
    }

    realized.push(realizedRow(tx, openDate, sellQty, costBasis, proceeds));
    t.qty -= sellQty;
    t.cost -= costBasis;
    if (t.qty <= 0) t.cost = 0;
  }
  return realized;
}

/* Aggregate realized lots into a tax breakdown for one fiscal year — mirrors
 * the backend's /tax/breakdown. rateApplied and carryForward are carried over
 * from the seed breakdown since offline mode has no tax-settings table. */
export function computeTaxBreakdown(realized, year, { rateApplied = 0.30, carryForward = 0 } = {}) {
  let shortTermGain = 0, longTermGain = 0, shortTermLoss = 0, longTermLoss = 0;
  for (const r of realized) {
    if (new Date(r.closeDate).getFullYear() !== year) continue;
    const g = r.gainMXN;
    if (r.kind === "short") {
      if (g >= 0) shortTermGain += g; else shortTermLoss += -g;
    } else {
      if (g >= 0) longTermGain += g; else longTermLoss += -g;
    }
  }
  return {
    year,
    rateApplied,
    shortTermGain: round2(shortTermGain),
    longTermGain: round2(longTermGain),
    shortTermLoss: round2(shortTermLoss),
    longTermLoss: round2(longTermLoss),
    carryForward,
  };
}
